libsel4camkes/docs/virtqueue.md

```diff
@@ -0,0 +1,133 @@
+    uses VirtQueueDrv drv;
+}
+
+component Foo2 {
+    uses VirtQueueDev dev;
+}
+
+assembly {
+    composition {
+        component VirtQueueInit vq_init;
+        component Foo bar;
+        component Foo2 baz;
+        
+        connection virtqueue_foo seL4VirtQueue(to vq_init.init, from bar.drv, from baz.dev);
+    }
+    
+    configuration {
+        bar.drv_id = 1;
+        baz.dev_id = 1;
+        bar.drv_shmem_size = 4096;
+        baz.dev_shmem_size = 4096;
+        virtqueue_foo.init_topology = [{"drv" : "bar.drv", "dev": "baz.dev" }];
+    }
+}
+```
+
+From the example above, the driver side of the virtqueue is `bar` and the
+device side is `baz`. The IDs of this particular virtqueue channel is 1 for
+both components. It is possible to have multiple channels for a component, but
+each channel's ID must be different. The memory size of the channel is 4096 and
+can also be changed, however, it must be page-aligned to 4096.
+
+Next, in the application code, the driver and device side of the channel must
+initialise their end of the channel by calling the appropriate function from
+the following:
+
+```c
+static inline int camkes_virtqueue_driver_init(virtqueue_driver_t *driver, unsigned int camkes_virtqueue_id) { ... }
+
+static inline int camkes_virtqueue_device_init(virtqueue_device_t *device, unsigned int camkes_virtqueue_id) { ... }
+```
+
+There is also a version of these functions which can return the underlying
+notification object that is backing the notification part of the virtqueues.
+
+```c
+int camkes_virtqueue_driver_init_with_recv(virtqueue_driver_t *driver, unsigned int camkes_virtqueue_id, 
+                                           seL4_CPtr *recv_notification, seL4_CPtr *recv_badge);
+                                           
+int camkes_virtqueue_device_init_with_recv(virtqueue_device_t *device, unsigned int camkes_virtqueue_id,
+                                           seL4_CPtr *recv_notification, seL4_CPtr *recv_badge);
+```
+
+The rest of the functions are essentially helper or convenience functions over
+the virtqueue interface as defined
+[here](https://github.com/SEL4PROJ/projects_libs/blob/master/libvirtqueue/include/virtqueue.h).
+
+### Driver
+
+When sending data over to the device side, it is preferred to use to these two functions.
+
+```c
+int camkes_virtqueue_driver_send_buffer(virtqueue_driver_t *vq, void *buffer, size_t size);
+
+int camkes_virtqueue_driver_scatter_send_buffer(virtqueue_driver_t *vq, void *buffer, size_t size);
+```
+
+When collecting buffers from the device side, the client must first obtain a
+handle to a used buffer ring via this virtqueue interface function:
+
+```c
+int virtqueue_get_used_buf(virtqueue_driver_t *vq, virtqueue_ring_object_t *robj, uint32_t *len);
+```
+
+It is then possible to gather each buffer in the used buffer ring and copy it into a larger buffer uisng the first function or iterate through each buffer in the ring using the second function:
```

Comment:
It should be 72 chars max everywhere.

Reply:
I don't think we have a decided value for markdown/documentation wrap. We have 120 for C and 100 for Python, so 72 sounds a bit small, but smaller than code wrap is justified for long text, I think.

It would probably make sense to add a markdown linter ([this one][0] seems widespread) to the style checks and with that we can encode a wrap value as well (default for the one linked above is 80, but easily adjusted).

[0]: https://github.com/DavidAnson/markdownlint

Reply:
I'm leaving the line wrap at 80 for now, because that is what most of the docs seem to be using. We can have a discussion about a global rule for this on a separate issue.